tests/unit/test_charm.py

```diff
@@ -0,0 +1,335 @@
+
+
+def test_start_sets_necessary_config(harness):
+    harness.add_relation(PEER, CHARM_KEY)
+    zk_rel_id = harness.add_relation(ZK, ZK)
+    harness.add_relation_unit(zk_rel_id, "zookeeper/0")
+    harness.update_relation_data(
+        zk_rel_id,
+        ZK,
+        {
+            "username": "relation-1",
+            "password": "mellon",
+            "endpoints": "123.123.123",
+            "chroot": "/kafka",
+            "uris": "123.123.123/kafka",
+            "tls": "disabled",
+        },
+    )
+
+    with patch("config.KafkaConfig.set_jaas_config") as patched_jaas, patch(
+        "config.KafkaConfig.set_server_properties"
+    ) as patched_properties:
+        harness.charm.on.start.emit()
+        patched_jaas.assert_called_once()
+        patched_properties.assert_called_once()
+
+
+def test_start_sets_auth_and_broker_creds_on_leader(harness):
+    peer_rel_id = harness.add_relation(PEER, CHARM_KEY)
+    zk_rel_id = harness.add_relation(ZK, ZK)
+    harness.add_relation_unit(zk_rel_id, "zookeeper/0")
+    harness.update_relation_data(
+        zk_rel_id,
+        ZK,
+        {
+            "username": "relation-1",
+            "password": "mellon",
+            "endpoints": "123.123.123",
+            "chroot": "/kafka",
+            "uris": "123.123.123/kafka",
+            "tls": "disabled",
+        },
+    )
+    harness.update_relation_data(peer_rel_id, CHARM_KEY, {"sync-password": "mellon"})
+
+    with patch("auth.KafkaAuth.add_user") as patched, patch(
+        "config.KafkaConfig.set_jaas_config"
+    ), patch("config.KafkaConfig.set_server_properties"), patch(
+        "charm.broker_active"
+    ) as patched_active:
+        patched_active.retry.wait = wait_none
+        harness.charm.on.start.emit()
+        patched.assert_not_called()
+
+        harness.set_leader(True)
+        harness.charm.on.start.emit()
+        patched.assert_called_once()
+
+
+def test_start_does_not_start_if_not_ready(harness):
+    peer_rel_id = harness.add_relation(PEER, CHARM_KEY)
+    zk_rel_id = harness.add_relation(ZK, ZK)
+    harness.add_relation_unit(zk_rel_id, "zookeeper/0")
+    harness.update_relation_data(
+        zk_rel_id,
+        ZK,
+        {
+            "username": "relation-1",
+            "password": "mellon",
+            "endpoints": "123.123.123",
+            "chroot": "/kafka",
+            "uris": "123.123.123/kafka",
+            "tls": "disabled",
+        },
+    )
+    harness.update_relation_data(peer_rel_id, CHARM_KEY, {"sync-password": "mellon"})
+
+    with patch("auth.KafkaAuth.add_user"), patch("config.KafkaConfig.set_jaas_config"), patch(
+        "config.KafkaConfig.set_server_properties"
+    ), patch(
+        "charm.KafkaCharm.ready_to_start", new_callable=PropertyMock, return_value=False
+    ), patch(
+        "snap.KafkaSnap.start_snap_service"
+    ) as patched_service:
+        harness.charm.on.start.emit()
+
+        patched_service.assert_not_called()
```

Comment:
might be nice to check if defer was called

Reply:
Very good check.

Reply:
I will however add a new test `if not ready, assert defer got called`